lib/navigation/preprocess_encoders.hpp

```diff
@@ -1,27 +1,96 @@
 #pragma once
+#include "consts.hpp"
 
+#include <cmath>
+
+#include <algorithm>
 #include <array>
 #include <cstdint>
+#include <optional>
 
+#include "core/logger.hpp"
 #include "core/types.hpp"
 
 namespace hyped::navigation {
+
 class EncodersPreprocessor {
  public:
-  EncodersPreprocessor();
+  EncodersPreprocessor(core::ILogger &logger);
 
-  core::EncoderData processData(const core::EncoderData encoder_data);
+  /**
+   * @brief process encoder data to deal with outliers and unreliable sensors
+   *
+   * @param encoder_data
+   * @return cleaned and reliable encoder data
+   */
+  std::optional<core::EncoderData> processData(const core::EncoderData &encoder_data);
 
  private:
-  core::EncoderData detectOutliers(const core::EncoderData encoder_data);
+  struct Statistics {
+    core::Float median;
+    core::Float upper_bound;
+    core::Float lower_bound;
+  };
+
+  std::optional<Statistics> getStatistics(const core::EncoderData &encoder_data) const;
+
+  /**
+   * @brief detects new outliers and replaces values from faulty sensors with
+   * the median; fails if not enough reliable data is available
+   *
+   * @param encoder_data to be sanitised
+   * @return consistent data with no outliers, if possible
+   */
+  std::optional<core::EncoderData> sanitise(const core::EncoderData &encoder_data);
+
+  /**
+   * @brief changes the value corresponding to the encoder sensor in the encoders_reliable_ array to
+   * false if the number of consecutive outliers exceeds the threshold value
+   *
+   * @return SensorChecks::kUnacceptable if more than 1 sensors have become unreliable and
+   * SensorChecks::kAcceptable otherwise
+   */
+  SensorChecks checkReliable();
 
-  void checkReliable(const core::EncoderData &encoder_data);
+  /**
+   * @brief calculates a specific quantile value from sorted input array
+   *
+   * @param reliable_data sorted array of values
+   * @param fraction in [0, 1] corresponding to the quantile (e.g. 0.25 for the 25th percentile)
+   * @return the quantile that we require
+   */
+  template<std::size_t N>
+  core::Float getSpecificQuantile(const std::array<std::uint32_t, N> &reliable_data,
+                                  const core::Float fraction) const
+  {
+    const core::Float theoretical_index = (num_reliable_encoders_ - 1) * fraction;
+    const std::size_t low_index         = static_cast<std::size_t>(std::floor(theoretical_index));
+    const std::size_t high_index        = static_cast<std::size_t>(std::ceil(theoretical_index));
```

Comment:
`auto` for these 2

Reply:
why do we need auto?

Reply:
it's clear from the `static_cast<std::size_t>` that the type is `std::size_t` so there is no need to duplicate it right next to it

Reply:
But if you directly static cast to a type, why then just call it `auto`? Surely you're as well using the type you just explicitly casted to?

Reply:
I think there is a discussion to be had on Monday about what exactly gets `auto`ed + various other minor standards. This is fine as it is for now